### DISCUSSION TITLE
how to set material "binding" with variable?

### DISCUSSION THREAD

**Discussion by wangzhaohao** [idaholab/moose#22869] [Q&A General] [answered]
hello, i want to set a PDE like this $\bigtriangledown (v_{p} \times P) = \frac{\partial P}{\partial t} $, in this, $v_{p}$ is a Function about location. and the week form is
 $<\varphi v_{p} P,\vec{n}> - (\bigtriangledown \varphi v_{p}P)-(\varphi, \frac{\partial P}{\partial t} )=0$ , 
but when i see the result， the moose solve as
 $<\varphi v_{p} P,\vec{n}> - v_{p}(\bigtriangledown \varphi P)-(\varphi, \frac{\partial P}{\partial t} )=0$ .
 my project is on my github. [https://github.com/wangzhaohao/restruction/blob/main/src/kernels/ConservativePorosity.C](url)
is there somthing i can do like comsol, in comsol, d(u+x, x) = ux+1, and pd(u+x, x) = 1. what can i do to tell moose $v_{p}$ is use inside. Do you have any suggestion? thanks for your reply.
Zhaohao Wang

**Comment by GiudGiud** [comment, marked as answer]
Hello

So v_p is a MOOSE Function, it is a pre-determined potentially spatially-dependent and potentially time-dependent profile that you do not have to solve for?
Or do you mean you want v_p to be a nonlinear variable that the solver should solve for, same as pressure?

Guillaume

**Reply by wangzhaohao** [reply]
v_p is a known function. i dont need to  solve

**Reply by GiudGiud** [reply]
and you want it to be inside the gradient?
I dont think we have an easy option to parse the gradient of that quantity for you, but you may write it explicitly in the kernel using the classic dab_dx = da_dx + db_dx

**Reply by wangzhaohao** [reply]
you mean is i take two part. and i see in heat conduction module the conductivity is set outside. there is we ignore divergence? why we dont need conside that.

**Reply by GiudGiud** [reply]
If it s constant spatially we do not need to consider it.
@WilkAndy might know

**Reply by leezw37** [reply]
> 

Hello, GiudGiud. What if i want v_p to be a nonlinear variable that the solver should solve for, same as pressure? I just break it  into two parts right now. Is there any magic way to simplify the process.

**Reply by WilkAndy** [reply]
I'm not exactly sure what is being asked here.  Is P a scalar variable, and v_p another scalar variable?  Is P a vector variable and v_p a vector-value function (is there a cross-product between v_p and P?)?   Or something else?   What is the difference between the two weak forms that you have written above?   And what does this mean: "d(u+x, x) = ux+1, and pd(u+x, x) = 1" ?    Sorry for my confusion - I can't help you without understand the question.

**Reply by wangzhaohao** [reply]
> I'm not exactly sure what is being asked here.  Is P a scalar variable, and v_p another scalar variable?  Is P a vector variable and v_p a vector-value function (is there a cross-product between v_p and P?)?   Or something else?   What is the difference between the two weak forms that you have written above?   And what does this mean: "d(u+x, x) = ux+1, and pd(u+x, x) = 1" ?    Sorry for my confusion - I can't help you without understand the question.

on my question v_p is a function which is konwn about location, and p is a variable. they are all about x. v_p is not product by P. the different about two week form is whether v_p is a function. if is a function, i think the div is acting on v_p and P. if is not a function, div is not acting on v_p only on P. 
my confusion point is how to set to let div acting   on v_p and P.

**Reply by WilkAndy** [reply]
> > I'm not exactly sure what is being asked here.  Is P a scalar variable, and v_p another scalar variable?  Is P a vector variable and v_p a vector-value function (is there a cross-product between v_p and P?)?   Or something else?   What is the difference between the two weak forms that you have written above?   And what does this mean: "d(u+x, x) = ux+1, and pd(u+x, x) = 1" ?    Sorry for my confusion - I can't help you without understand the question.
> 
> on my question v_p is a function which is konwn about location, and p is a variable. they are all about x. v_p is not product by P. the different about two week form is whether v_p is a function. if is a function, i think the div is acting on v_p and P. if is not a function, div is not acting on v_p only on P. my confusion point is how to set to let div acting on v_p and P.

I'm beginning to understand.

Probably you're trying to solve the advection equation (the green equation in https://en.wikipedia.org/wiki/Advection ).   If that is true, then the weak form is encoded in the https://mooseframework.inl.gov/source/kernels/TimeDerivative.html and https://mooseframework.inl.gov/source/kernels/ConservativeAdvection.html Kernels (along with any `BCs` that are appropriate for you).   You don't need to code any C++ to solve the advection equation using MOOSE: just use those two in-built Kernels.

If you've trying to solve du/dt + v.grad(u) = 0, where u is your `Variable` and `v` is a known vector field, then first note that this is not the advection equation.  I suggest writing this as du/dt + div(vu) - u div(v) = 0.  Then the first two terms are the advection equation, and are encoded as described in the last paragraph, while the final term can be described using https://mooseframework.inl.gov/source/kernels/CoefReaction.html (but div(v) has to be a constant) or https://mooseframework.inl.gov/source/kernels/MatReaction.html (you'd have to create a Material version of your Function)

**Reply by wangzhaohao** [reply]
Thanks，i got it
